sinabs/backend/dynapcnn/nir_graph_extractor.py

```diff
@@ -0,0 +1,401 @@
+            if old_idx in remapped_nodes
+        }
+
+        # Update entry nodes based on new node indices
+        self._entry_nodes = {
+            remapped_nodes[old_idx]
+            for old_idx in self._entry_nodes
+            if old_idx in remapped_nodes
+        }
+
+        # Update io-shapes based on new node indices
+        self._nodes_io_shapes = {
+            remapped_nodes[old_idx]: shape
+            for old_idx, shape in self._nodes_io_shapes.items()
+            if old_idx in remapped_nodes
+        }
+
+        # Update sinabs module map based on new node indices
+        self._modules_map = {
+            remapped_nodes[old_idx]: module
+            for old_idx, module in self._modules_map.items()
+            if old_idx in remapped_nodes
+        }
+
+    def get_node_io_shapes(self, node: int) -> Tuple[torch.Size, torch.Size]:
+        """Returns the I/O tensors' shapes of `node`.
+
+        Returns
+        ----------
+        - input shape (torch.Size): shape of the input tensor to `node`.
+        - output shape (torch.Size): shape of the output tensor from `node`.
+        """
+        return (
+            self._nodes_io_shapes[node]["input"],
+            self._nodes_io_shapes[node]["output"],
+        )
+
+    ####################################################### Pivate Methods #######################################################
+
+    def _get_edges_from_nir(
+        self, nir_graph: nirtorch.graph.Graph
+    ) -> Tuple[List[Tuple[int, int]], Dict[str, int], List[int]]:
+        """Standardize the representation of `nirtorch.graph.Graph` into a list of edges (`Tuple[int, int]`) where each node in `nir_graph` is represented by an interger (with the source node starting as `0`).
+
+        Parameters
+        ----------
+        - nir_graph (nirtorch.graph.Graph): a NIR graph representation of `spiking_model`.
+
+        Returns
+        ----------
+        - edges (set): tuples describing the connections between layers in `spiking_model`.
+        - name_2_indx_map (dict): `key` is the original variable name for a layer in `spiking_model` and `value is an integer representing the layer in a standard format.
+        - entry_nodes (set): IDs of nodes acting as entry points for the network (i.e., receiving external input).
+        """
+        edges = set()
+        name_2_indx_map = {}
+
+        # TODO maybe make sure the input node from nir always gets assined `0`.
+
+        for src_node in nir_graph.node_list:
+            # Make sure current node is in `name_2_indx_map`
+            if src_node.name not in name_2_indx_map:
+                # Assign unique index by taking current length of `name_2_indx_map`
+                name_2_indx_map[src_node.name] = len(name_2_indx_map)
+
+            for trg_node in src_node.outgoing_nodes:
+                # Make sure all targets of current node are in `name_2_indx_map`
+                if trg_node.name not in name_2_indx_map:
```

Comment:
Why is this necessary? Shouldn't we get all nodes from already from iterating over `nir_graph.node_list`?

Reply:
Ah got it. To be able to include the target node index in the edge.
I simplified this a bit. @Willian-Girao Please check.